### ISSUE TITLE
TorchScript to MNN 失败

### ISSUE THREAD

**Issue by shinobuwz** [alibaba/MNN#2410]
使用如下代码测试，可以正常导出MNN模型
~~~python
from torch import nn
import torch

# 定义模型
class MyModel(nn.Module):
    def __init__(self):
        super(MyModel, self).__init__()
        self.a = 1
        self.b = 2
        self.linear = nn.Linear(784, 10)

    def forward(self, x):
        return self.linear(x)


# 创建模型实例
model = MyModel()

# 将模型转换为eval模式
model.eval()

# 创建一个样例输入
example_input = torch.rand(1, 784)

# 使用trace函数将模型转换为图形表示
traced_model = torch.jit.trace(model, example_input)

# 保存导出模型
traced_model.save('exported_model.pt')
~~~

转换结果
~~~bash
The device support i8sdot:0, support fp16:0, support i8mm: 0
Start to Convert Other Model Format To MNN Model...
Start to Optimize the MNN Net...
Reshape error: 1 -> 0
inputTensors : [ x.1, ]
outputTensors: [ 6, ]
Converted Success!
~~~

但是转换我自己的模型时，使用相同的方式导出，一直无法转化成功
~~~python
import os
import models
import torch
import config

def torch2onnx(cfg):
    # cfg = config.v9
    sample_rate = cfg['data_config']['fs']
    frame_len = int(sample_rate * cfg['ms'])
    frame_shift = frame_len//2
    frame_num = 1
    batch_size = 1

    device = torch.device('cpu')
    myModule = models.myModel(cfg['net_config'])

    model_name = cfg['model_name']
    checkpoint_path = cfg['checkpoint_path'] + '_multi'
    model_path = os.path.join(checkpoint_path, f"{model_name}.pt")

    # if os.path.exists(model_path):
    #     state_dict = torch.load(model_path, map_location=torch.device(device))
    #     new_state_dict = {}
    #     for key, value in state_dict.items():
    #         new_key = key.replace('module.', '')
    #         new_state_dict[new_key] = value
    #     myModule.load_state_dict(new_state_dict)
    #     myModule.to(device)
    #     print("loading model done!")
    # else:
    #     print("no trained model")
    #     exit()

    input_shape = (batch_size,frame_shift + 1, frame_num, 1)
    noisy_wav = torch.randn(*input_shape).to(device)
    input_data = (noisy_wav, noisy_wav)

    myModule.eval()

    model_script = torch.jit.script(myModule)
    export_jit_file = os.path.join(checkpoint_path, f"{model_name}_jit.pt")
    model_script.save(export_jit_file)

    model_trace = torch.jit.trace(myModule, input_data)
    export_jit_file2 = os.path.join(checkpoint_path, f"{model_name}_jit2.pt")
    model_trace.save(export_jit_file2)


if __name__ == '__main__':
    import argparse
    import config
    parse = argparse.ArgumentParser()
    parse.add_argument("--use_config", type=str, default='v9')
    args = parse.parse_args()

    version = args.use_config
    use_config = getattr(config, version)
    torch2onnx(use_config)
~~~

转换结果
~~~bash
The device support i8sdot:0, support fp16:0, support i8mm: 0
Start to Convert Other Model Format To MNN Model...
[ERROR] TorchScript model can't load. Please using `torch.jit.script` or `torch.jit.trace` save model.
[ERROR] Convert error, please check your file format.
~~~

平台：linux

**Comment by jxt1234** [comment]
呃，看着好像没什么问题，有能复现的完整工程么？

**Comment by shinobuwz** [comment]
> 呃，看着好像没什么问题，有能复现的完整工程么？

~~~python
import os
import torch
from torch import nn
import torch.nn.functional as F

class Conv_Block(nn.Module):
    def __init__(self, in_channels,
                 out_channels,
                 kernel_size,
                 stride):
        super(Conv_Block, self).__init__()
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.kernel_size = kernel_size
        self.stride = stride
        self.conv_block = nn.Sequential(nn.Conv2d(self.in_channels, self.out_channels, self.kernel_size, self.stride),
                                        nn.BatchNorm2d(self.out_channels),
                                        nn.PReLU(self.out_channels))

    def forward(self, x):
        # x.shape: [B, C, F, T]
        out = self.conv_block(x)
        return out

class ConvTr_Block(nn.Module):
    def __init__(self, in_channels,
                 out_channels,
                 kernel_size,
                 stride,
                 padding=(0, 0),
                 norm_act=True):
        super(ConvTr_Block, self).__init__()
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.kernel_size = kernel_size
        self.stride = stride
        self.padding = padding
        self.norm_act = norm_act
        if self.norm_act:
            self.convtr_block = nn.Sequential(
                nn.ConvTranspose2d(self.in_channels, self.out_channels, self.kernel_size, self.stride, self.padding),
                nn.BatchNorm2d(self.out_channels),
                nn.PReLU(self.out_channels))
        else:
            self.convtr_block = nn.Sequential(
                nn.ConvTranspose2d(self.in_channels, self.out_channels, self.kernel_size, self.stride, self.padding),
                nn.BatchNorm2d(self.out_channels))

    def forward(self, x):
        # x.shape: [B, C, F, T]
        out = self.convtr_block(x)
        return out

class ProjRNN_ONNX(nn.Module):
    def __init__(self, input_size, hidden_size, rnn_type='LSTM', dropout=0, bidirectional=False):
        super(ProjRNN_ONNX, self).__init__()

        self.input_size = input_size
        self.rnn_type = rnn_type
        self.hidden_size = hidden_size
        self.num_direction = int(bidirectional) + 1

        self.rnn = getattr(nn, rnn_type)(input_size, hidden_size, 1, dropout=dropout, batch_first=True,
                                         bidirectional=bidirectional)
        self.proj = nn.Linear(hidden_size * self.num_direction, input_size)

    def forward(self, input, h_in=None):

        output = input
        if h_in is not None:
            rnn_output, h_out = self.rnn(output, h_in)
            proj_output = self.proj(rnn_output.contiguous().view(-1, rnn_output.shape[2])).view(output.shape)
            return proj_output, h_out
        else:
            rnn_output, _ = self.rnn(output)
            proj_output = self.proj(rnn_output.contiguous().view(-1, rnn_output.shape[2])).view(output.shape)
            return proj_output

class DualPathRNN_ONNX(nn.Module):
    def __init__(self, rnn_type, input_size, hidden_size, width,
                 dropout=0, num_layers=2, bidirectional=True):
        super(DualPathRNN_ONNX, self).__init__()

        self.rnn_type = rnn_type
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.width = width

        # dual-path RNN
        self.row_rnn = nn.ModuleList([])
        self.col_rnn = nn.ModuleList([])
        self.row_norm = nn.ModuleList([])
        self.col_norm = nn.ModuleList([])
        for i in range(num_layers):
            self.row_rnn.append(ProjRNN_ONNX(self.input_size, self.hidden_size // 2, self.rnn_type, dropout,
                                        bidirectional=True))
            self.col_rnn.append(
                ProjRNN_ONNX(self.input_size, self.hidden_size, self.rnn_type, dropout, bidirectional=bidirectional))
            self.row_norm.append(nn.LayerNorm([self.width, self.input_size], eps=1e-8))
            self.col_norm.append(nn.LayerNorm([self.width, self.hidden_size], eps=1e-8))

    def forward(self, input, h_in_list):

        batch_size, channels, width, frame_num = input.shape  # B, C, W, T

        output = input.permute(0, 3, 2, 1).contiguous()  # B, T, W, C
        hidden_out = []
        for index, (row_rnn, row_norm, col_rnn, col_norm) in enumerate(zip(self.row_rnn, self.row_norm, self.col_rnn, self.col_norm)):
            row_input = output.contiguous().view(batch_size * frame_num, width, channels)  # B*T, W, C
            row_output = row_rnn(row_input)  # B*T, W, C
            row_output = row_output.view(batch_size, frame_num, width, channels).contiguous()  # B, T, W, C
            row_output = row_norm(row_output)
            output = output + row_output

            h_in = h_in_list[index]
            col_input = output.permute(0, 2, 1, 3).contiguous().view(batch_size * width, frame_num, channels)  # B*W, T, C
            col_output, h_out = col_rnn(col_input, h_in)  # B*W, T, C
            col_output = col_output.view(batch_size, width, frame_num, channels).permute(0, 2, 1, 3).contiguous()  # B, T, W, C
            col_output = col_norm(col_output)
            output = output + col_output
            hidden_out.append(h_out)

        output = output.permute(0, 3, 2, 1).contiguous()  # B, C, W, T

        return output, hidden_out

class DPCRN_ONNX(nn.Module):
    def __init__(self,
                 conf):
        super(DPCRN_ONNX, self).__init__()

        conv_block = conf['conv_block']
        dprnn_block = conf['dprnn_block']
        convTr_block = conf['convTr_block']

        self.n_fft = conf['n_fft']
        self.hop_length = conf['hop_length']
        self.conv_pad = conf['conv_pad']
        self.convTr_pad = conf['convTr_pad']

        if conf['spec_norm'] == 'BatchNorm':
            self.spec_norm = nn.BatchNorm2d(3)
        elif conf['spec_norm'] == 'InstanceNorm':
            self.spec_norm = nn.InstanceNorm2d(3, affine=True)
        self.encoder = nn.ModuleList([])
        for conv in conv_block:
            self.encoder.append(
                Conv_Block(conv['Cin'], conv['Cout'], kernel_size=conv['kernel'], stride=conv['stride']))

        self.dprnn_block = DualPathRNN_ONNX(dprnn_block['rnn_type'], dprnn_block['in_size'], dprnn_block['hidden_size'],
                                       width=dprnn_block['width'],
                                       dropout=dprnn_block['dropout'], num_layers=dprnn_block['num_layers'],
                                       bidirectional=dprnn_block['bidirectional'])

        self.decoder = nn.ModuleList([])
        for convTr in convTr_block:
            self.decoder.append(
                ConvTr_Block(convTr['Cin'], convTr['Cout'], kernel_size=convTr['kernel'], stride=convTr['stride'],
                             norm_act=convTr['norm']))

        self.mag_mask = ConvTr_Block(convTr_block[-1]['Cin'], 1, kernel_size=convTr_block[-1]['kernel'],
                                     stride=convTr_block[-1]['stride'], norm_act=False)

        self.act = conf['mag_act']
        self.mask_bn = nn.BatchNorm2d(1)
        if conf['mag_act'] == 'sigmoid':
            self.mag_act = nn.Sigmoid()
        elif conf['mag_act'] == 'softplus':
            self.mag_act = nn.Softplus()

    def mk_mask_mag(self, noisy_real, noisy_imag, mag_mask):
        '''
          magnitude mask
        '''
        enh_mag_real = noisy_real * mag_mask
        enh_mag_imag = noisy_imag * mag_mask
        return enh_mag_real, enh_mag_imag

    def mk_mask_pha(self, enh_mag_real, enh_mag_imag, pha_cos, pha_sin):
        '''
        phase mask
        '''
        enh_real = enh_mag_real * pha_cos - enh_mag_imag * pha_sin
        enh_imag = enh_mag_real * pha_sin + enh_mag_imag * pha_cos

        return enh_real, enh_imag

    def forward(self, noisy_real, noisy_imag, h_in0, h_in1):
        noisy_mag = torch.sqrt(torch.square(noisy_real) + torch.square(noisy_imag) + 1e-9)
        noisy_log_mag = 2 * torch.log(noisy_mag)

        spec_input = torch.cat([noisy_real, noisy_imag, noisy_log_mag], dim=-1)
        spec_input = spec_input.permute(0, 3, 1, 2).contiguous()  # [B, 3, F, T]
        norm_spec = self.spec_norm(spec_input)
        noisy_real = torch.squeeze(noisy_real, dim=-1)
        noisy_imag = torch.squeeze(noisy_imag, dim=-1)

        conv_in = norm_spec
        conv_out_list = []
        for idx, encoder in enumerate(self.encoder):
            pad_in = F.pad(conv_in, self.conv_pad[idx], "constant", 0.)
            conv_in = encoder(pad_in)
            conv_out_list.append(conv_in)
        rnn_in = conv_in

        rnn_out, h_out = self.dprnn_block(rnn_in, [h_in0, h_in1])

        convTr_in = rnn_out
        for idx, decoder in enumerate(self.decoder):
            depad_in = F.pad(torch.cat([convTr_in, conv_out_list[-(idx + 1)]], dim=1), (0, 0, 0, 0), "constant", 0.)
            convTr_in = decoder(depad_in)[:, :, :self.convTr_pad[idx], :]

        mag_mask = self.mag_mask(depad_in)[:, :, :self.convTr_pad[idx], :]
        mask = convTr_in

        # get magnitude mask
        if self.act == 'sigmoid':
            mag_mask = self.mag_act(self.mask_bn(mag_mask))[:, 0, :, :] * 1.2
        elif self.act == 'softplus':
            mag_mask = self.mag_act(self.mask_bn(mag_mask))[:, 0, :, :]
        # get phase mask
        phase_square = torch.sqrt(torch.square(mask[:, 0, :, :]) + torch.square(mask[:, 1, :, :]) + 1e-9)
        phase_sin = mask[:, 1, :, :] / phase_square
        phase_cos = mask[:, 0, :, :] / phase_square

        enh_mag_real, enh_mag_imag = self.mk_mask_mag(noisy_real, noisy_imag, mag_mask)
        enh_real, enh_imag = self.mk_mask_pha(enh_mag_real, enh_mag_imag, phase_cos, phase_sin)
        return enh_real, enh_imag, h_out[0], h_out[1]

net_config =  {
    'use_init': True,
    'n_fft': 1280,
    'hop_length': 640,
    'spec_norm': 'BatchNorm',
    'mag_act': 'softplus',
    'conv_block': [
        {'Cin': 3, 'Cout': 32, 'kernel': (5, 1), 'stride': (2, 1)},
        {'Cin': 32, 'Cout': 32, 'kernel': (3, 1), 'stride': (2, 1)},
        {'Cin': 32, 'Cout': 32, 'kernel': (3, 1), 'stride': (2, 1)},
        {'Cin': 32, 'Cout': 64, 'kernel': (3, 1), 'stride': (1, 1)},
        {'Cin': 64, 'Cout': 64, 'kernel': (3, 1), 'stride': (1, 1)},
    ],
    'conv_pad': [
        (0, 0, 0, 2),
        (0, 0, 0, 1),
        (0, 0, 0, 1),
        (0, 0, 1, 1),
        (0, 0, 1, 1),
    ],
    'dprnn_block': {
        'rnn_type': 'GRU',
        'in_size': 64,
        'hidden_size': 64,
        'width': 80,
        'dropout': 0,
        'num_layers': 2,
        'bidirectional': False
    },
    'convTr_pad': [
        -2,
        -2,
        -1,
        -1,
        -2,
    ],
    'convTr_block': [
        {'Cin': 128, 'Cout': 64, 'kernel': (3, 1), 'stride': (1, 1), 'norm': True},
        {'Cin': 128, 'Cout': 32, 'kernel': (3, 1), 'stride': (1, 1), 'norm': True},
        {'Cin': 64, 'Cout': 32, 'kernel': (3, 1), 'stride': (2, 1), 'norm': True},
        {'Cin': 64, 'Cout': 32, 'kernel': (3, 1), 'stride': (2, 1), 'norm': True},
        {'Cin': 64, 'Cout': 2, 'kernel': (5, 1), 'stride': (2, 1), 'norm': False},
    ]
}

def torch_export(cfg):

    frame_len = 1280
    frame_shift = frame_len//2
    frame_num = 1
    batch_size = 1
    in_size = net_config['dprnn_block']['width']
    hid_size = net_config['dprnn_block']['hidden_size']

    device = torch.device('cpu')
    myModule = DPCRN_ONNX(net_config)

    input_shape = (batch_size,frame_shift + 1, frame_num, 1)
    noisy_wav = torch.randn(*input_shape).to(device)
    h_in = torch.zeros((1, in_size, hid_size)).to(device)
    input_data = (noisy_wav, noisy_wav, h_in, h_in)

    myModule.eval()

    # model_script = torch.jit.script(myModule)
    # export_jit_file = "jit_script.pt"
    # model_script.save(export_jit_file)

    model_trace = torch.jit.trace(myModule, input_data)
    export_jit_file2 = "jit_trace.pt"
    model_trace.save(export_jit_file2)


if __name__ == '__main__':

    torch_export(net_config)
~~~
这里是可复现的文件，导出的模型用MNN最新的master在linux上的编译转换工具无法转换

**Comment by shinobuwz** [comment]
> 呃，看着好像没什么问题，有能复现的完整工程么？

虽然转ONNX再转MNN可以成功，但是转出来的模型算子太碎了。。。有工具能整合简化一下也行
